.github/workflows/nubots.yaml

```diff
@@ -0,0 +1,179 @@
+# Continuous Integration tests
+name: NUbots CI
+
+# Controls when the action will run.
+on:
+  # Triggers the workflow on push or pull request events but only for the main branch
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
+
+# A workflow run is made up of one or more jobs that can run sequentially or in parallel
+jobs:
+  # Build the docker image
+  build_docker:
+    name: "Build docker image"
+
+    # The type of runner that the job will run on
+    runs-on: ubuntu-20.04
+
+    # We output the image tag that we create so the other jobs can use it
+    outputs:
+      image: ${{ steps.image_output.outputs.image }}
+
+    # Steps represent a sequence of tasks that will be executed as part of the job
+    steps:
+      # Cancels previously running jobs for this workflow
+      - name: Cancel Workflow Action
+        uses: styfle/cancel-workflow-action@0.9.1
+
+      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
+      - name: Checkout Code
+        uses: actions/checkout@v2
+
+      # Setup docker buildx
+      - name: 🐳 Set up docker buildx 🐳
+        uses: docker/setup-buildx-action@v1
+
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_PASSWORD }}
+
+      # Build the docker image
+      - name: 🐳 Build the docker image 🐳
+        uses: docker/build-push-action@v2
+        with:
+          pull: true
+          tags: "nubots/nubots:${{ hashFiles('docker/**') }}"
+          file: docker/Dockerfile
+          context: docker
+          build-args: |
+            platform=generic
+          push: true
+          cache-from: type=registry,ref=nubots/nubots:generic
+          cache-to: type=inline
+
+      - id: image_output
+        name: Output the created image
+        run: echo "::set-output name=image::nubots/nubots:${{ hashFiles('docker/**') }}"
+
+  # Build the codebase
+  build_nubots:
+    name: "Build and test NUbots"
+
+    # The type of runner that the job will run on
+    runs-on: ubuntu-20.04
+    needs: build_docker
```

Comment:
Does this mean we will build the generic image and push it to dockerhub before we build the code?

Reply:
I believe so. @TrentHouliston was telling me that each PR will indeed have its own image pushed to dockerhub. I think that happens at the end of `build_docker` stage.

Reply:
Why push a image on PR? is that so later jobs can use the pushed image?

Reply:
Yessir

Reply:
If we want to keep this behaviour then we can add a step to this PR to delete those pushed images

Reply:
Did you want to add that to this PR rather than follow up? There will be follow-up PRs for not building nusight or nubots if either hasn't changed, and trying to cache C++ builds.
Personally, I am keen to avoid scope creep on this PR, but if you reckon it's worth waiting on this PR, I'll have a crack at doing it

Reply:
It seems like a natural action that should happen to he honest. This way, from the moment this PR gets merged, we wont have any "mess" left that needs a manual clean up later.

Reply:
Added the functionality in `dockerhub-cleanup.yaml` workflow :+1: